feat(atc): support chained container placement strategies.

evanchaoli opened the pull request:
## What does this PR accomplish?

Currently if a container placement strategy results multiple candidate workers, then it will randomly pick up one from the candidates.

Say, if `volume-locality` strategy results 2 candidate workers, why choose the one that has less containers or tasks?

With PR implements a strategy chain, so that if one strategy results multiple candidate workers, then next strategy can be applied.

## Changes proposed by this PR:

Organizes multiple strategies as a chain.

## Notes to reviewer:

Please review from the design perspective first. If you are ok with the design, I'll add more tests for chained strategies. This enhancement should not break any existing strategy.

## Release Note

* Enhanced container placement strategy to support chained strategies, for example `CONCOURSE_CONTAINER_PLACEMENT_STRATEGY=volume-locality,fewest-build-containers`

## Contributor Checklist
<!--
Most of the PRs should have the following added to them,
this doesn't apply to all PRs, so it is helpful to tell us what you did.
-->
- [ ] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [ ] [Signed] all commits
- [ ] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

xtreme-sameer-vohra commented on atc/worker/placement.go:
```diff
@@ -1,31 +1,104 @@
+			return nil, nil
+		}
+	}
+	if len(workers) == 1 {
```
Nvm, I can see this needs to be where it is in case the limit active tasks strategy is being used.

evanchaoli replied:
Yes, in the loop, when there is only 1 worker left, next strategy should still be applied, which may further filter out workers.

xtreme-sameer-vohra commented:
Thanks for the PR @evanchaoli 
This looks great.

cc @kcmannem who had dreamed this up many 🌔s ago

evanchaoli commented:
@xtreme-sameer-vohra Thanks for your review. I'll add more test cases (I think I mentioned that in "Notes to reviewer"), just haven't got a chance to get back on this PR yet.

xtreme-sameer-vohra commented:
> @xtreme-sameer-vohra Thanks for your review. I'll add more test cases (I think I mentioned that in "Notes to reviewer"), just haven't got a chance to get back on this PR yet.

Sounds good. Feel free to include the changes here https://github.com/concourse/concourse/pull/6233

evanchaoli commented on atc/worker/placement.go:
```diff
@@ -1,31 +1,104 @@
+				return nil, errors.New("max-active-tasks-per-worker must be greater or equal than 0")
+			}
+			cps.nodes = append(cps.nodes, newLimitActiveTasksPlacementStrategy(opts.MaxActiveTasksPerWorker))
+		default:
```
I just copied the original code, so that is the current Concourse's behavior. But I agree explicit is better. Fixed.

evanchaoli commented:
@xtreme-sameer-vohra I took the test case you added from #6233. Also addressed the review comments. So I think this PR is ready to ship. Please take a look again.

evanchaoli commented on atc/worker/placement.go:
```diff
@@ -1,31 +1,104 @@
+			cps.nodes = append(cps.nodes, newFewestBuildContainersPlacementStrategy())
+		case "limit-active-tasks":
+			if opts.MaxActiveTasksPerWorker < 0 {
+				return nil, errors.New("max-active-tasks-per-worker must be greater or equal than 0")
```
Added a test.

evanchaoli commented:
@muntac Can you please help move forward this PR?

aoldershaw commented on atc/worker/placement.go:
```diff
@@ -1,31 +1,106 @@
 )
 
+type ContainerPlacementStrategyOptions struct {
+	ContainerPlacementStrategy string `long:"container-placement-strategy" default:"volume-locality" choice:"volume-locality" choice:"random" choice:"fewest-build-containers" choice:"limit-active-tasks" description:"Method by which a worker is selected during container placement. Multiple methods can be chained by +, e.g. volume-locality+fewest-build-containers"`
```
Perhaps it'd be better to make this a `[]string` (and rename the field to `ContainerPlacementStrategies`, but keep the flag as is).

You'd then configure `concourse web` with:

```
--container-placement-strategy=volume-locality --container-placement-strategy=fewest-build-containers
```

or

```
CONCOURSE_CONTAINER_PLACEMENT_STRATEGY=volume-locality,fewest-build-containers
```

That way, it's consistent with other flags that accept multiple values

evanchaoli replied:
Make sense. I just didn't realize `[]string` can be used. 😅

aoldershaw commented on atc/worker/placement.go:
```diff
@@ -123,31 +194,27 @@ func (strategy *LimitActiveTasksPlacementStrategy) Choose(logger lager.Logger, w
-func (strategy *RandomPlacementStrategy) Choose(logger lager.Logger, workers []Worker, spec ContainerSpec) (Worker, error) {
-	return workers[strategy.rand.Intn(len(workers))], nil
+func (strategy *RandomPlacementStrategyNode) Choose(logger lager.Logger, workers []Worker, spec ContainerSpec) ([]Worker, error) {
+	return []Worker{workers[strategy.rand.Intn(len(workers))]}, nil
```
Perhaps this should return all workers and let `containerPlacementStrategy.Choose()` deal with selecting at random.

Though there's no good reason to do it, if you specified `CONCOURSE_CONTAINER_PLACEMENT_STRATEGY= random,volume-locality`, the `volume-locality` strategy would have no effect, since we already filtered the list down to a single worker.

So, `random` is in a bit of a weird spot. It might be more valuable to view these as `XYZPlacementStrategyNodes` as filters - in which case, `RandomPlacementStrategyNode` is really an `IdentityFilter` or something like that - i.e. it just returns the full list of workers that it's provided.

evanchaoli replied:
This makes sense. I thought to keep `random` strategy for backward-compatibility. Now, I think we can just deprecate `random` strategy, and states that in the option description.

aoldershaw replied:
I don't know if we should get rid of random altogether - it just doesn't totally make sense with chaining. I think just making the RandomPlacementStrategyNode return the full list of workers is sufficient, though, since it won't lead to any unexpected behaviour with chaining

evanchaoli replied:
I don't remove "random" in this PR. I just added option description to state deprecation.

aoldershaw commented on atc/worker/placement.go:
```diff
@@ -49,26 +123,24 @@ func (strategy *VolumeLocalityPlacementStrategy) Choose(logger lager.Logger, wor
 
-type FewestBuildContainersPlacementStrategy struct {
+type FewestBuildContainersPlacementStrategyNode struct {
 	rand *rand.Rand
```
Looks like we can get rid of `rand` here, as well as in `LimitActiveTasksPlacementStrategyNode` and possibly `RandomPlacementStrategyNode` (see my comment there)

evanchaoli replied:
Good catch.

evanchaoli commented:
@aoldershaw Thanks for your review and comments. I have addressed your comments, please take a look again.

aoldershaw commented on atc/worker/placement.go:
```diff
@@ -1,31 +1,101 @@
 )
 
+type ContainerPlacementStrategyOptions struct {
+	ContainerPlacementStrategy []string `long:"container-placement-strategy" default:"volume-locality" choice:"volume-locality" choice:"random" choice:"fewest-build-containers" choice:"limit-active-tasks" description:"Method by which a worker is selected during container placement. If multiple methods are specified, they will be applied in order. Random strategy is deprecated, it is still accepted,  ...
```
I think we should actually get rid of the deprecation notice - although it's true that no strategy is the same as a random strategy, I think it's not possible to specify no strategy, given that there's a default of `volume-locality`.

So, I don't think we'll even want to deprecate `random`, since operators should be allowed to use completely random placement if they so choose - but maybe we can mention that chaining `random` is redundant

evanchaoli replied:
OK, make sense.

aoldershaw reviewed (approved):
Thanks! LGTM
